tst/testall.g

```diff
@@ -4,7 +4,13 @@
 # This file runs package tests. It is also referenced in the package
 # metadata in PackageInfo.g.
 #
-LoadPackage( "typeset" );
+LoadPackage( "ctbllib" );
+LoadPackage( "digraphs" );
+
+# Reread package instead of loading due to digraphs functionality
+# potentially not being loaded.
+RereadPackage( "typeset", "init.g");
+RereadPackage( "typeset", "read.g");
```

Comment:
Why not just this? Note that testall.g should load `typeset` in any case...
```suggestion
LoadPackage( "ctbllib" );
LoadPackage( "digraphs" );

LoadPackage( "typeset" );
```

Reply:
Agree with @fingolfin. Same for me - I see participating/mentioned/review requested, the number of other notifications is unmanageable.

Reply:
For some reason only using `LoadPackage("typeset");`  in testall.g will cause errors if the following workflow is followed (assuming `gap -A` is run):

```
gap> LoadPackage("typeset":OnlyNeeded);
gap> TestPackage("typeset":OnlyNeeded);
```

I believe there might be some issues in an initial load prior to testing super-ceding the loading in `testall.g` - since the digraphs functionality is only loaded in `init.g` if the package is marked for loading this could cause error. It seems that the second call to LoadPackage doesn't run through `init.g` and recognise that `digraphs` is now loaded.

```
 ┌───────┐   GAP 4.12.0 of 2022-08-18
 │  GAP  │   https://www.gap-system.org
 └───────┘   Architecture: x86_64-pc-linux-gnu-default64-kv8
 Configuration:  gmp 6.2.1, GASMAN
 Loading the library and packages ...
 Packages:   GAPDoc 1.6.6, PrimGrp 3.4.2, SmallGrp 1.5, TransGrp 3.6.3
 Try '??help' for help. See also '?copyright', '?cite' and '?authors'
gap> LoadPackage("typeset":OnlyNeeded);
─────────────────────────────────────────────────────────────────────────────
Loading  typeset 1.2.0 (Automatic typesetting framework for common GAP objects, with LaTeX generation)
by Zachariah Newbery (https://zachnewbery.com).
Homepage: https://gap-packages.github.io/typeset/
Report issues at https://github.com/gap-packages/typeset/issues
─────────────────────────────────────────────────────────────────────────────
true
gap> TestPackage("typeset":OnlyNeeded);
... Digraphs dependencies ...
─────────────────────────────────────────────────────────────────────────────
Loading  Digraphs 1.5.3 (Digraphs - Methods for digraphs)
... ...
─────────────────────────────────────────────────────────────────────────────
Architecture: x86_64-pc-linux-gnu-default64-kv8

testing: /home/parrot/Downloads/gap-4.12.0/pkg/typeset/tst/latex/core.tst
      45 ms (41 ms GC) and 568KB allocated for latex/core.tst
testing: /home/parrot/Downloads/gap-4.12.0/pkg/typeset/tst/latex/ctbls.tst
      96 ms (37 ms GC) and 15.3MB allocated for latex/ctbls.tst
testing: /home/parrot/Downloads/gap-4.12.0/pkg/typeset/tst/latex/digraphs.tst
########> Diff in /home/parrot/Downloads/gap-4.12.0/pkg/typeset/tst/latex\
/digraphs.tst:6
# Input is:
Typeset(g);
# Expected output:
\begin{center}
\begin{tikzpicture}[>=latex',line join=bevel,]
\begin{dot2tex}[dot,tikz,codeonly,styleonly]
digraph hgn{
node [shape=circle]
1
2
1 -> 1
1 -> 2
2 -> 2
2 -> 1
}
\end{dot2tex}
\end{tikzpicture}
\end{center}
# But found:
<immutable digraph with 2 vertices, 4 edges>
########
########> Diff in /home/parrot/Downloads/gap-4.12.0/pkg/typeset/tst/latex\
/digraphs.tst:23
# Input is:
Typeset(g2);
# Expected output:
\begin{center}
\begin{tikzpicture}[>=latex',line join=bevel,]
\begin{dot2tex}[dot,tikz,codeonly,styleonly]
digraph hgn{
node [shape=circle]
1
2
3
4
5
1 -> 1
1 -> 2
2 -> 2
2 -> 3
3 -> 3
3 -> 4
4 -> 4
4 -> 2
5 -> 4
5 -> 1
}
\end{dot2tex}
\end{tikzpicture}
\end{center}
# But found:
<immutable digraph with 5 vertices, 10 edges>
########
########> Diff in /home/parrot/Downloads/gap-4.12.0/pkg/typeset/tst/latex\
/digraphs.tst:56
# Input is:
PositionSublist(s, "\\node")<>fail;
# Expected output:
true
# But found:
false
########
      38 ms (37 ms GC) and 98.3KB allocated for latex/digraphs.tst
testing: /home/parrot/Downloads/gap-4.12.0/pkg/typeset/tst/latex/generators.ts\
t
      49 ms (33 ms GC) and 2.03MB allocated for latex/generators.tst
testing: /home/parrot/Downloads/gap-4.12.0/pkg/typeset/tst/latex/polynomials.t\
st
      39 ms (34 ms GC) and 582KB allocated for latex/polynomials.tst
testing: /home/parrot/Downloads/gap-4.12.0/pkg/typeset/tst/latex/structdesc.ts\
t
      35 ms (33 ms GC) and 82.3KB allocated for latex/structdesc.tst
testing: /home/parrot/Downloads/gap-4.12.0/pkg/typeset/tst/mathml/core.tst
      36 ms (34 ms GC) and 207KB allocated for mathml/core.tst
-----------------------------------
total       338 ms (249 ms GC) and 18.9MB allocated
              3 failures in 1 of 7 files

#I  Errors detected while testing
```

Reply:
When you call `LoadPackage` for a package that is already loaded, it has no effect. You cannot load a package second time. So there is no surprise here.

You are not supposed to call `LoadPackage` before `TestPackage`. Rather your `TestFile` should load the packages, possibly after loading other packages are optional dependencies but are nevertheless required for the test.

And that's exactly why my suggested changes does.